### PR TITLE
Refactor scan and patch steps in parallel and provide report after patching

### PR DESCRIPTION
**Pull request by g-iannelli**

### CODE REVIEW

File: CVEs/patch_images_with_copacetic.sh [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    info "Update patching report for $image_to_patch"
+    image_to_patch_hash=$(jq -r '.Metadata.ImageID' < "$TRIVY_SCAN_OUTPUT_FILE")
+    for FIXED_CVE in ${FIXED_CVES[@]}
+    do
+      DOCKER_LABELS="--label io.sighup.secured.${FIXED_CVE}=fixed ${DOCKER_LABELS}"
+      jq -r \
+      --arg cve "$FIXED_CVE" \
+      --arg image_to_patch "$image_to_patch" \
+      --arg image_to_patch_hash "$image_to_patch_hash" \
+      --arg image_patched "$image_to_patch-patched" \
```
Comment (ralgozino) [inline CVEs/patch_images_with_copacetic.sh:105]:
```suggestion
      --arg image_patched "${image_to_patch}-patched" \
```

File: CVEs/patch_images_with_copacetic.sh [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+      --arg image_patched_hash "$image_patched_hash" \
+      '[try .Results[].Vulnerabilities[] | select(.VulnerabilityID==$cve)][0] | "|" + $image_to_patch + " | " + $image_to_patch_hash + " | " + .VulnerabilityID + " | " +  .Severity +  " | " + .Title + " | " + $image_patched + "|" + $image_patched_hash + "|"' < "$TRIVY_SCAN_OUTPUT_FILE">> "$PATCH_REPORT_OUTPUT_FILE"
+    done
+    info "Tag $image_to_patch-patched as secured image: $secured_image"
+    echo "FROM $image_to_patch-patched" | DOCKER_BUILDKIT=0 docker build \
```
Comment (ralgozino) [inline CVEs/patch_images_with_copacetic.sh:110]:
```suggestion
    info "Tag ${image_to_patch}-patched as secured image: $secured_image"
    echo "FROM ${image_to_patch}-patched" | DOCKER_BUILDKIT=0 docker build \
```

File: CVEs/patch_images_with_copacetic.sh [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      ${DOCKER_LABELS} \
+      --label io.sighup.secured.image.created="$(date -u +"%Y-%m-%dT%H:%M:%S.%3NZ")" \
+      --label io.sighup.secured.image.from.hash="$image_to_patch_hash" \
+      -t "$secured_image" \
+      -f - "$DOCKERFILE_OUTPUT_DIR" &> /dev/null
+    secured_labeled_image_hash=$(docker inspect "$secured_image" --format '{{.Id}}')
+    sed -i'.unsecured' s#"$image_to_patch-patched"#"$secured_image"# "$PATCH_REPORT_OUTPUT_FILE"
```
Comment (ralgozino) [inline CVEs/patch_images_with_copacetic.sh:117]:
```suggestion
    sed -i'.unsecured' s#"${image_to_patch}-patched"#"$secured_image"# "$PATCH_REPORT_OUTPUT_FILE"
```

File: CVEs/patch_images_with_copacetic.sh [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    sed -i'.unsecured' s#"$image_patched_hash"#"$secured_labeled_image_hash"# "$PATCH_REPORT_OUTPUT_FILE"
+    rm "$PATCH_REPORT_OUTPUT_FILE.unsecured"
+    info "Push secure image: $secured_image"
+    [[ $DRY_RUN -eq 0 ]] && docker push "$secured_image"
+    info "Cleanup $image_to_patch-patched"
+    buildctl --addr tcp://127.0.0.1:8888 prune
+    docker rmi -f "$image_to_patch-patched"
```
Comment (ralgozino) [inline CVEs/patch_images_with_copacetic.sh:124]:
```suggestion
    info "Cleanup ${image_to_patch}-patched"
    buildctl --addr tcp://127.0.0.1:8888 prune
    docker rmi -f "${image_to_patch}-patched"
```

File: CVEs/scan_vuln.sh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+mkdir -p "$TRIVY_SCAN_OUTPUT_DIR"
+
+for image in $IMAGE_LIST; do
+  TRIVY_SCAN_OUTPUT_FILE="$TRIVY_SCAN_OUTPUT_DIR/scan-${image//[:\/]/_}.json"
+
+  info "Scan $image for CVEs"
+  if ! trivy image --skip-db-update --skip-java-db-update --scanners vuln --no-progress --output "$TRIVY_SCAN_OUTPUT_FILE" --format json --severity CRITICAL "$image"
+  then
+    echo "$image | ERROR PROCESSING! " >> "${SCAN_ERROR_OUTPUT_FILE}"
```
Comment (ralgozino) [inline CVEs/scan_vuln.sh:60]: should we log also here to stdout with error or warning?

**Review by ralgozino — CHANGES_REQUESTED** [review]

**Review by ralgozino — CHANGES_REQUESTED** [review]
The How to got outdated with the latest changes, it needs to be worked on a little.

I left some very small suggestions in the code, nothing critical, only stye.

File: CVEs/patch_images_with_copacetic.sh [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+  info "Start buildkitd instance for COPA"
+  if [ -n "$DOCKER_CONFIG" ]
+  then
+    docker_config_extra_args="-v ${DOCKER_CONFIG}:/root/.docker"
+  fi
+  docker run --detach --rm --privileged $docker_config_extra_args -p 127.0.0.1:8888:8888/tcp --name buildkitd --entrypoint buildkitd registry.sighup.io/fury-secured/moby/buildkit:v0.16.0 --addr tcp://0.0.0.0:8888 # --platform linux/amd64
+fi
+
+mkdir -p "$TRIVY_SCAN_OUTPUT_DIR" "$COPA_PATCH_OUTPUT_DIR" "$DOCKERFILE_OUTPUT_DIR" "$LOG_OUTPUT_DIR"
+echo -n "" > "${PATCH_ERROR_OUTPUT_FILE}"
```
[the remark this answers is not in the document]
Comment (g-iannelli) [reply CVEs/patch_images_with_copacetic.sh:52]: On MacOs touch will not clean the file content.

```
$ echo -n "" > example_file
$ echo "hello world" >> example_file

$ cat example_file
hello world

$ touch example_file

$ cat example_file
hello world
```

[the remark this answers is not in the document]
Comment (ralgozino) [reply CVEs/patch_images_with_copacetic.sh:52]: you are absolutely right

File: CVEs/Makefile [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+kustomize-build-all:
+	# to remove KFD unused cluster-autoscaler images
+	KFD_VERSION=$(KFD_VERSION)
+	echo "" > $(KFD_VERSION)/built.yaml
+	for kustomize in `find $(KFD_VERSION)/.furyctl/sighup/vendor/modules \
+		\( -name "kustomization.y*ml" \) \
+	  	-not -path "*cluster-autoscaler/*/kustomization.yaml" \
+		-and -not -path "*examples*" \
+		-or -path "*/cluster-autoscaler/$${KFD_VERSION%.*}*/kustomization.yaml"`; do \
+			`find $(KFD_VERSION)/.furyctl/bin -name kustomize -type f` build `dirname $$kustomize` >> $(KFD_VERSION)/built.yaml; \
```
[the remark this answers is not in the document]
Comment (g-iannelli) [reply CVEs/Makefile:58]: The awk search is quite fast. It is performed in about 4 seconds. Maybe it will not improve the speed of the task.

**Review by ralgozino — APPROVED** [review]
code-wise LGTM.

**Review by ralgozino — APPROVED** [review]

**Pull request merged**